# Revisão Python: 0 to Hero

### 1. Escreva um programa que verifica se um número é par ou ímpar.

In [10]:
numero = float(input('Digite um número e descubra se ele é impar ou par.\n'))
if numero%2 == 0:
    print("Par")
else:
    print("Impar")

Impar


### 2. Crie um programa que calcula a média de três números inseridos pelo usuário.

In [2]:
n1 = float(input('\nDigite um número\n'))
n2 = float(input('\nDigite outro número\n'))
n3 = float(input('\nDigite o último número\n'))
media = n1 + n2 + n3/3
print(f'A média entre os três números é {media}')

A média entre os três números é 4.0


### 3. Faça um programa que imprime os números de 1 a 10 usando um loop for.

In [5]:
for i in range(1,11):
    print(i)

1
2
3
4
5
6
7
8
9
10


### 4. Escreva um programa que solicita ao usuário um número e imprime todos os números pares até ele.

In [5]:
while True:
    user_input = input('\nDigite um número: ')
    
    
    if user_input.isdigit():
        user_number = int(user_input)
        
        for i in range(user_number):
            if i % 2 == 0:
                print(i)
        
        break 
    else:
        print("Entrada inválida. Por favor, digite um número inteiro positivo válido.")  

Entrada inválida. Por favor, digite um número inteiro positivo válido.
0
2
4
6
8
10


### 5. Implemente um programa que imprime a série de Fibonacci até o décimo termo.

In [4]:
def fibonacci(n):
    fib_sequence = [0, 1]
    for i in range(2, n):
        next_fib = fib_sequence[i-1] + fib_sequence[i-2]
        fib_sequence.append(next_fib)
    return fib_sequence

n = 12  
fib_numbers = fibonacci(n)
print(fib_numbers)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]


### 6. Crie um programa que calcule a soma dos números de 1 a 100 usando um loop while.

In [1]:
num = 100
i = 0
while i < num:
    print(i)
    i+= 1

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99


### 7. Faça um programa que peça ao usuário um número e verifique se é positivo, negativo ou zero.

In [ ]:
while True:
    try:
        number = float(input('Digite um número: '))
        
        result = ("Positivo" if number > 0 else "Zero") if number == 0 else "Negativo"
        print("O número é:", result)
        
        break  
    
    except ValueError:
        print("Entrada inválida. Por favor, digite um número válido.")

### 8. Escreva um programa que imprime os múltiplos de 3 de 1 a 50.

In [4]:
for i in range(50):
    if i%3 == 0:
        print(i)

0
3
6
9
12
15
18
21
24
27
30
33
36
39
42
45
48


### 9. Crie um programa que calcule a potência de um número informado pelo usuário usando um loop for.

In [6]:
base = float(input("Digite a base: "))
expoente = int(input("Digite o expoente: "))
resultado = 1

for _ in range(expoente):
    resultado *= base

print(f"{base} elevado a {expoente} é igual a {resultado}")

5.0 elevado a 123 é igual a 9.403954806578305e+85


### 10. Crie um programa que calcule o fatorial de um número inserido pelo usuário.

In [13]:
def calcular_fatorial(numero):
    if numero == 0 or numero == 1:
        return 1
    else:
        fatorial = 1
        for i in range(2, numero + 1):
            fatorial *= i
        return fatorial
   
try:
    numero_inserido = int(input("Digite um número para calcular o fatorial: "))
    if numero_inserido < 0:
        print("O fatorial não está definido para números negativos.")
    else:
        resultado = calcular_fatorial(numero_inserido)
        print(f"O fatorial de {numero_inserido} é {resultado}")
except ValueError:
    print("Por favor, insira um número inteiro válido.")

O fatorial de 3 é 6


### 11. Crie um programa que leia um arquivo de texto e conte quantas palavras ele contém.

In [17]:
num_of_worlds = 0
with open(r'sla.txt', 'r') as file:
    data = file.read()
    lines = data.split()
    num_of_worlds += len(lines)
    print(num_of_worlds)

15


### 12. Escreva um programa que leia um arquivo e imprima o conteúdo em ordem reversa.

In [8]:
with open(r'sla.txt', 'r') as file:
    linhas = file.readlines()
    linhas_reversas = reversed(linhas)

    print("Conteúdo do arquivo em ordem reversa:")
    for linha in linhas_reversas:
            print(linha.strip())


Conteúdo do arquivo em ordem reversa:

It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.
when an unknown printer took a galley of type and scrambled it to make a type specimen book.
Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
Lorem Ipsum is simply dummy text of the printing and typesetting industry.


### 13. Faça um programa que copie o conteúdo de um arquivo para outro, ao contrário.

### 14. Implemente um programa que conte quantas vezes uma determinada palavra aparece em um arquivo.

### 15. Escreva um programa que renomeie um arquivo.

### 16. Implemente um programa que crie um novo diretório e mova para lá todos os arquivos da pasta atual que possuem a extensão especificada pelo usuário.

### 17. Crie um programa que compare dois arquivos informados pelo usuário e imprima as linhas em que eles diferem.

### 18. Crie uma função que verifique se um número é primo e retorne "Verdadeiro" ou "Falso".

### 19. Implemente uma função que retorne o máximo de três números.

### 20. Escreva uma função que inverta uma string e retorne o resultado.

### 21. Faça uma função que receba uma lista de números e retorne a soma deles.

### 22. Crie uma função que verifique se uma string é um palíndromo.

### 23. Faça uma função que retorne uma lista de todos os divisores inteiros de um número.

### 24. Crie uma função que converta temperatura em Celsius para Fahrenheit.

### 25. Crie um programa que simule um jogo de adivinhação. O programa escolherá um número aleatório entre 1 e 100, e o usuário terá que adivinhar qual é o número. O programa deve fornecer dicas se o palpite do usuário for muito alto ou muito baixo.

### 26. Implemente um programa que calcule a soma dos dígitos de um número inteiro inserido pelo usuário. Por exemplo, se o número for 123, a saída deve ser 6 (1 + 2 + 3).

### 27. Faça um programa que receba um número inteiro e exiba a tabuada desse número, de 1 a 10.

### 28. Faça um programa que leia um número inteiro positivo e exiba todos os números primos até esse número.

### 29. Escreva um programa que calcule a sequência de Collatz para um número inserido pelo usuário e exiba cada iteração na tela. A sequência de Collatz é gerada da seguinte forma: se o número for par, divida-o por 2; se for ímpar, multiplique-o por 3 e adicione 1. Repita o processo até chegar ao número 1.

### 30. Faça um programa que leia um arquivo de texto e crie um novo arquivo com todas as palavras em ordem alfabética.

### 31. Faça um programa que leia um arquivo de texto e substitua todas as ocorrências de uma determinada palavra por outra palavra inserida pelo usuário.

### 32. Escreva um programa que leia um arquivo de texto contendo linhas de código em Python e identifique todas as funções definidas no arquivo, exibindo seus nomes.

### 33. Crie uma função que calcule o n-ésimo termo da sequência de Fibonacci.

### 34. Crie uma função que calcule o Máximo Divisor Comum (MDC) de dois números usando o algoritmo de Euclides.

### 35. Escreva uma função que verifique se um número é um quadrado perfeito e retorne "SIM" ou "NÃO".

### 36. Crie uma função que receba uma lista de palavras e retorne a palavra mais longa.

### 37. Crie um programa que simule o jogo "Pedra, Papel e Tesoura". O usuário escolhe uma das opções, o computador sorteia outra e o programa determina o vencedor.

### 38. Implemente um programa que calcule o valor aproximado de π usando a fórmula de Leibniz. *

### 39. Crie um programa que gere uma lista de números primos usando a técnica do Crivo de Eratóstenes. *

### 40. Escreva um programa que resolva a Torre de Hanoi para um número de discos inserido pelo usuário.

### 41. Escreva um programa que converta um número inteiro em sua representação binária usando um loop.

### 42. Implemente uma função que calcule a potência de um número usando recursão. A função deve aceitar dois argumentos: base e expoente.

### 43. Faça uma função que retorne uma lista contendo todos os números perfeitos em um intervalo dado. Um número perfeito é aquele cuja soma de seus divisores próprios é igual a ele mesmo.

### 44. Implemente uma função que calcule a média móvel de uma lista de números para um dado tamanho de janela.

### 45. Escreva uma função que gere todas as combinações possíveis de um conjunto de elementos dados.

### 46. Faça uma função que receba um número inteiro e retorne uma lista de seus dígitos em ordem inversa.

### 47. Implemente uma função que encontre o menor múltiplo comum (MMC) de dois números.

### 48. Crie uma função que calcule a raiz quadrada de um número usando o método de Newton-Raphson.

### 49. Você foi contratado para desenvolver um programa de controle de estoque para uma loja. O programa deve permitir ao usuário adicionar produtos ao estoque, realizar vendas e verificar o estoque atual. Cada produto possui um nome, preço unitário e quantidade em estoque.

Seu programa deve ter as seguintes funcionalidades:

a) Adicionar um produto ao estoque, solicitando ao usuário o nome, preço unitário e quantidade inicial.

b) Realizar uma venda, solicitando ao usuário o nome do produto e a quantidade vendida. O programa deve verificar se há estoque suficiente e, se sim, calcular o total da venda e atualizar o estoque.

c) Exibir o estoque atual, listando o nome, preço unitário e quantidade de cada produto.

### 50. Você foi contratado para criar um sistema de gerenciamento de tarefas. O programa deve permitir que o usuário crie, visualize, marque como concluída e liste tarefas. Cada tarefa possui um título, uma descrição e um status (concluída ou não).

Seu programa deve ter as seguintes funcionalidades:

a) Adicionar uma nova tarefa à lista, solicitando ao usuário o título e a descrição.

b) Marcar uma tarefa como concluída, exibindo a lista de tarefas e permitindo ao usuário escolher qual tarefa marcar.

c) Exibir a lista de tarefas, mostrando o título e o status de cada uma.

d) Listar apenas as tarefas concluídas.

e) Listar apenas as tarefas não concluídas.

Além disso, implemente uma função que calcule a porcentagem de tarefas concluídas em relação ao total de tarefas.